### PR TITLE
Requirements update

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

This PR adds a resumable iterable dataset (`LightningIterableDataset`)

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: src/lightning/data/datasets/env.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+    def __init__(self, world_size: int, global_rank: int):
+        self.world_size = world_size
+        self.global_rank = global_rank
+
+    @classmethod
+    def detect(cls) -> "DistributedEnv":
+        """Tries to automatically detect the distributed environment paramters.
+
+        Note:
```
Comment (Borda) [inline src/lightning/data/datasets/env.py:23]:
```suggestion
        .. note::
```

File: src/lightning/pytorch/callbacks/spike.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        trainer: pl.Trainer,
+        pl_module: pl.LightningModule,
+        outputs: Union[torch.Tensor, Mapping[str, torch.Tensor]],
+        batch: Any,
+        batch_idx: int,
+    ) -> None:
+        if isinstance(outputs, torch.Tensor):
+            loss = outputs.detach()
+        elif isinstance(outputs, Mapping):
+            loss = outputs["loss"].detach()
```
Comment (ethanwharris) [inline src/lightning/pytorch/callbacks/spike.py:24]: Maybe nicer error if `loss` key is missing

**Review by ethanwharris — APPROVED** [review]
LGTM just small things

File: src/lightning/data/datasets/env.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+    @property
+    def num_shards(self) -> int:
+        """Returns the total number of shards.
+
+        Note:
+            This may not be accurate in a non-dataloader-worker process like the main training process
+            as it doesn't necessarily know about the number of dataloader workers.
+        """
+        assert self.worker_env is not None
```
[the remark this answers is not in the document]
Comment (justusschock) [reply src/lightning/data/datasets/env.py:121]: ignoring you :D

File: src/lightning/data/datasets/base.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    def __init__(self, backend: str = "local"):
+        self.backend = self._init_backend(backend=backend)
+
+        assert isinstance(self.backend, _DatasetBackend)
+
+    def _init_backend(self, backend: str) -> _DatasetBackend:
+        """Picks the correct backend handler."""
+        if backend == "s3":
+            return S3DatasetBackend()
```
[the remark this answers is not in the document]
Comment (justusschock) [reply src/lightning/data/datasets/base.py:24]: the base class is just protected so that nobody tries to use it directly. This was previously somewhere else, just moved it to this class for code sharing. Left everything else as is, won't touch it now :D

File: src/lightning/pytorch/callbacks/spike.py (hunk 2 above)
Comment (justusschock) [reply src/lightning/pytorch/callbacks/spike.py:24]: from a training_step it can't be missing. Otherwise optimization (specifically backward) would fail earlier.

File: src/lightning/fabric/utilities/spike.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        self.running_mean = Running(MeanMetric(dist_sync_on_step=False, sync_on_compute=False), window=window)
+        self.mode = mode
+        self.warmup = warmup
+        self.atol = atol
+        self.rtol = rtol
+        self.bad_batches: List[int] = []
+        self.exclude_batches_path = exclude_batches_path
+
+    @torch.no_grad()
+    def on_train_batch_end(self, fabric: "_BooleanReducer", loss: torch.Tensor, batch: Any, batch_idx: int) -> None:
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/spike.py:41]: Why is fabric annotated like this? I see it everywhere.

File: src/lightning/fabric/utilities/spike.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
+import json
+import operator
+import os
+import warnings
+from typing import Any, Dict, List, Literal, Optional, Protocol, runtime_checkable, Union
+
+import torch
+from lightning_utilities.core.imports import compare_version
+
+_TORCHMETRICS_GREATER_EQUAL_1_0_0 = compare_version("torchmetrics", operator.ge, "1.0.0")
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/spike.py:10]: Was there a problem with the regular RequirementCache? We should probably use it.

File: src/lightning/fabric/utilities/spike.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+if _TORCHMETRICS_GREATER_EQUAL_1_0_0:
+    from torchmetrics.aggregation import MeanMetric
+    from torchmetrics.wrappers import Running
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/spike.py:13]: I suggest moving this to local import following the practices we use everywhere. To avoid #12786

File: src/lightning/fabric/utilities/spike.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    def _check_rtol(self, val_a: Union[float, torch.Tensor], val_b: Union[float, torch.Tensor]) -> bool:
+        return (self.rtol is None) or bool(abs(val_a - val_b) >= abs(self.rtol * val_b))
+
+    def _is_better(self, diff_val: torch.Tensor) -> bool:
+        if self.mode == "min":
+            return bool((diff_val <= 0.0).all())
+        if self.mode == "max":
+            return bool((diff_val >= 0).all())
+
+        raise ValueError(f"invalid mode. Has to be min or max, found {self.mode}")
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/spike.py:103]:
```suggestion
        raise ValueError(f"Invalid mode. Has to be min or max, found {self.mode}")
```

File: src/lightning/fabric/utilities/spike.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+import torch
+from lightning_utilities.core.imports import compare_version
+
+_TORCHMETRICS_GREATER_EQUAL_1_0_0 = compare_version("torchmetrics", operator.ge, "1.0.0")
+if _TORCHMETRICS_GREATER_EQUAL_1_0_0:
+    from torchmetrics.aggregation import MeanMetric
+    from torchmetrics.wrappers import Running
+
+
+class SpikeDetection:
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/spike.py:16]: I propose to add a docstring, explaining what each argument does

File: src/lightning/pytorch/callbacks/spike.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
+import os
```
Comment (awaelchli) [inline src/lightning/pytorch/callbacks/spike.py:1]: This probably needs to be added to the callbacks `__init__.py`

**Review by awaelchli — COMMENTED** [review]
I suggest that the spike detection be a separate PR. These two features can be reviewed and land independently, right?

File: src/lightning/fabric/utilities/spike.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+        if not str(self.exclude_batches_path).endswith(".json"):
+            self.exclude_batches_path = os.path.join(self.exclude_batches_path, "skip_batches.json")
+
+        is_spike = bool(batch_idx >= self.warmup and self.is_spike(loss))
+
+        # While spike-detection happens on a per-rank level, we need to fail all ranks if any rank detected a spike
+        is_spike_global = fabric.strategy.reduce_boolean_decision(is_spike, all=False)
+
+        print(is_spike_global, fabric.global_rank)
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/spike.py:56]: Left from debugging?

Comment (awaelchli) [reply src/lightning/fabric/utilities/spike.py:56]:
```suggestion
```

File: src/lightning/fabric/utilities/spike.py (hunk 5 above)
Comment (justusschock) [reply src/lightning/fabric/utilities/spike.py:41]: just so that I can also pass the trainer to it. We don't really need everything from fabric there, just reducing a boolean

File: src/lightning/fabric/utilities/spike.py (hunk 9 above)
Comment (justusschock) [reply src/lightning/fabric/utilities/spike.py]: done in https://github.com/Lightning-AI/lightning/pull/18014

File: src/lightning/fabric/utilities/spike.py (hunk 7 above)
Comment (justusschock) [reply src/lightning/fabric/utilities/spike.py]: done in https://github.com/Lightning-AI/lightning/pull/18014

File: src/lightning/fabric/utilities/spike.py (hunk 5 above)
Comment (justusschock) [reply src/lightning/fabric/utilities/spike.py]: changed in https://github.com/Lightning-AI/lightning/pull/18014

File: src/lightning/pytorch/callbacks/spike.py (hunk 10 above)
Comment (justusschock) [reply src/lightning/pytorch/callbacks/spike.py]: done in https://github.com/Lightning-AI/lightning/pull/18014

**Review by nohalon — APPROVED** [review]
lgtm

File: src/lightning/data/datasets/base.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+
+class _Dataset(TorchDataset):
+    """Base dataset class for streaming data from a cloud storage.
+
+    Args:
+        backend: storage location of the data_source. current options are "s3" or "local"
+    """
+
+    def __init__(self, backend: str = "local"):
```
Comment (awaelchli) [inline src/lightning/data/datasets/base.py:16]:
```suggestion
    def __init__(self, backend: Literal["local", "s3"] = "local"):
```
as per docstring

File: src/lightning/data/datasets/iterable.py [hunk 13]
```diff
@@ -0,0 +1,447 @@
...
+            On resume from a state-dict, we always skip currently started chunks as these would make the data-order
+            impossible to determine with sharding. Upon resuming from a point, where a new chunk would be started
+            anyways, nothing is skipped.
+
+        Note:
+            Order of data is only guaranteed when resuming with the same distributed settings and the same number of
+            workers. Everything else leads to different sharding and therefore results in different data order.
+    """
+
+    # TODO: backend docstring
```
Comment (awaelchli) [inline src/lightning/data/datasets/iterable.py:104]: Resolve the TODO now or is it meant for future?

File: src/lightning/data/datasets/env.py [hunk 14]
```diff
@@ -0,0 +1,147 @@
+from typing import Optional
+
+import torch
+from torch.utils.data import get_worker_info
+
+
+class DistributedEnv:
```
Comment (awaelchli) [inline src/lightning/data/datasets/env.py:7]: I suggest making these helper classes protected. I expect they need to be changed in the future, so it would be easier to do if we didn't expose them to users (yet). wdyt?

File: src/lightning/data/datasets/iterable.py [hunk 15]
```diff
@@ -0,0 +1,447 @@
...
+
+    def prepare_chunk(self, chunk: Any) -> None:
+        """Prepares a single chunk before it is actually loaded. This could e.g. download the actual file from s3.
+
+        Args:
+            chunk: the chunk data to prepare.
+        """
+
+    @abstractmethod
+    def load_chunk(self, chunk: Any) -> Any:
```
Comment (awaelchli) [inline src/lightning/data/datasets/iterable.py:165]: I suggest to put the two abstract methods first after the init

File: src/lightning/data/datasets/iterable.py [hunk 16]
```diff
@@ -0,0 +1,447 @@
...
+        backend: str = "local",
+    ):
+        _StatefulIterableDataset.__init__(self)
+        _Dataset.__init__(self, backend=backend)
+
+        chunks = [_Chunk(c, chunk_size=chunk_size) for c in chunks]
+        if env is None:
+            # must detect distributed env here since distributed is not initialized in worker processes with ddp spawn
+            # can't detect worker env here since workers not yet initialized
+            env = Environment(DistributedEnv.detect(), None)
```
Comment (awaelchli) [inline src/lightning/data/datasets/iterable.py:124]: I understand this only partially. Why can we assume the dist env is initialized here? The creation of the dataset might happen in the main process before processes have been spawned and dist is init. I suggest to delay the creation of env entirely to `__iter__`. It looks like `self._env` is not needed during the dataset init.

File: src/lightning/data/datasets/iterable.py [hunk 17]
```diff
@@ -0,0 +1,447 @@
...
+        """Prepares a single chunk before it is actually loaded. This could e.g. download the actual file from s3.
+
+        Args:
+            chunk: the chunk data to prepare.
+        """
+
+    @abstractmethod
+    def load_chunk(self, chunk: Any) -> Any:
+        """Loads a single chunk into memory. This could e.g. mean loading the file that has previously been
+        downloaded from s3.
```
Comment (awaelchli) [inline src/lightning/data/datasets/iterable.py:167]: Since the user needs to implement this hook, maybe an imperative style docstring could be better? E.g., "Implement this to load a single chunk into memory ...". Same suggestion below

File: src/lightning/data/datasets/iterable.py [hunk 18]
```diff
@@ -0,0 +1,447 @@
...
+
+            sample = self.load_sample_from_chunk(
+                self._curr_loaded_chunks[curr_loaded_chunk_idx]._chunk_data,
+                self._curr_loaded_chunks[curr_loaded_chunk_idx].index_permutations[remainder],
+            )
+
+            return sample
+
+        finally:
+            self._curr_sample_index += 1
```
Comment (awaelchli) [inline src/lightning/data/datasets/iterable.py:238]: Could you remind me why is it correct to increment the index when failing on the current one? Wouldn't we want to "resume" on the last incomplete sample?

File: src/lightning/data/datasets/iterable.py [hunk 19]
```diff
@@ -0,0 +1,447 @@
...
+        }
+
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+        """Loads a previous state dict to resume for it's state.
+
+        Args:
+            state_dict: the previous state-dict containing internal indices and random number generator states.
+
+        Note:
+            some of the changes only take effect when creating a new iterator
```
Comment (awaelchli) [inline src/lightning/data/datasets/iterable.py:280]:
```suggestion
            Some of the changes only take effect when creating a new iterator
```

File: src/lightning/data/datasets/iterable.py [hunk 20]
```diff
@@ -0,0 +1,447 @@
...
+                chunk = _Chunk(None, chunk_size=self._chunk_size)
+                chunk.shuffle(generator=self._generator)
+        # shuffle samples within each chunk
+        elif self._shuffle:
+            chunks = [c.shuffle(generator=self._generator) for c in chunks]
+
+        return chunks
+
+
+class DataLoader(_DataLoader):
```
Comment (awaelchli) [inline src/lightning/data/datasets/iterable.py:401]: The user might not know they have to use this dataloader. And it will probably work without notice if you use the regular dataloader from torch, right? Will we communicate this through docs?

**Review by awaelchli — APPROVED** [review]
Minor nits and questions. Great work!

File: src/lightning/data/datasets/iterable.py (hunk 16 above)
Comment (justusschock) [reply src/lightning/data/datasets/iterable.py:124]: I had a problem initializing it in `__iter__` since the torch.distributed isn't initialized in the worker processes.

File: src/lightning/data/datasets/iterable.py (hunk 18 above)
Comment (justusschock) [reply src/lightning/data/datasets/iterable.py:238]: We always skip to the next full chunk. I can change the `finally` to do it immediately after the sample is loaded.

File: src/lightning/data/datasets/iterable.py (hunk 20 above)
Comment (justusschock) [reply src/lightning/data/datasets/iterable.py:401]: You can still use the dataloader for training, just resuming won't work unless you manually pass in the correct values for the seen samples and num_workers
